crates/oxc_linter_plugin/src/plugin.rs

```diff
@@ -0,0 +1,615 @@
+    pub fail: Vec<SingleTest>,
+}
+
+#[derive(Deserialize, Clone, Debug)]
+pub struct SingleTest {
+    pub relative_path: Vec<String>,
+    pub code: String,
+}
+
+#[derive(Debug)]
+pub struct LinterPlugin {
+    pub(super) rules: Vec<InputQuery>,
+    schema: &'static Schema,
+}
+
+pub enum RulesToRun {
+    All,
+    Only(String),
+}
+
+#[derive(Debug, Error, Diagnostic)]
+#[error(transparent)]
+pub struct ParseError(serde_yaml::Error);
+
+#[derive(Debug, Error, Diagnostic)]
+pub enum ErrorFromLinterPlugin {
+    #[error("{0}")]
+    PluginGenerated(String, String, #[label("{1}")] Span),
+    #[error("{error_message}")]
+    Trustfall {
+        error_message: String,
+        #[source_code]
+        query_source: Arc<NamedSource>,
+        #[label = "This query failed."]
+        query_span: SourceSpan,
+    },
+    #[error(transparent)]
+    Ignore(ignore::Error),
+    #[error(transparent)]
+    ReadFile(std::io::Error),
+    #[error("Failed to parse file at path: {0}")]
+    QueryParse(PathBuf, #[related] Vec<ParseError>),
+    #[error(
+        "Expected span_start and span_end to be List of Int or Int, instead got\nspan_start = {span_start}\nspan_end = {span_end}"
+    )]
+    WrongTypeForSpanStartSpanEnd {
+        span_start: String,
+        span_end: String,
+        #[source_code]
+        query_source: Arc<NamedSource>,
+        #[label = "This query failed."]
+        query_span: SourceSpan,
+    },
+}
+
+impl LinterPlugin {
+    pub fn new(schema: &'static Schema, queries_path: &PathBuf) -> oxc_diagnostics::Result<Self> {
+        let mut deserialized_queries = vec![];
+
+        for dir_entry_found_maybe in Walk::new(queries_path) {
+            let dir_entry_found = dir_entry_found_maybe.map_err(ErrorFromLinterPlugin::Ignore)?;
+
+            let pathbuf = dir_entry_found.path().to_path_buf();
+
+            if pathbuf.extension().map_or(false, |ext| ext.eq_ignore_ascii_case("yml")) {
+                let text = fs::read_to_string(&pathbuf).map_err(ErrorFromLinterPlugin::ReadFile)?;
+
+                let mut deserialized =
+                    serde_yaml::from_str::<InputQuery>(&text).map_err(|err| {
+                        ErrorFromLinterPlugin::QueryParse(pathbuf.clone(), vec![ParseError(err)])
+                    })?;
+                deserialized.path = pathbuf;
+                deserialized_queries.push(deserialized);
+            }
+        }
+
+        Ok(Self { rules: deserialized_queries, schema })
+    }
+
+    #[allow(clippy::too_many_lines)]
+    pub fn run_plugin_rules(
+        &self,
+        ctx: &mut LintContext,
+        plugin: &InputQuery,
+        adapter: &Adapter<'_>,
+        mv8: &mini_v8::MiniV8,
+    ) -> oxc_diagnostics::Result<()> {
+        let arc_adapter = Arc::new(adapter);
+        let query_source =
+            Arc::new(NamedSource::new(plugin.path.to_string_lossy(), plugin.query.clone()));
+        // NOTE: the 0 is technically wrong, but it's the right file which is enough
```

Comment:
Maybe add some more context here? Why is it enough that it's in the right file?

Reply:
It is now clear to me that this was not clear at all. Basically what this does is make a pseudo-file that is just the query, (so we don't have to reparse the whole file to get the span of the real query in the real file). So it's enough to say the right file because when they ctrl-click the file name in their cmdline, they will be brought to the right yaml file, but the span row will be incorrect since when we made the span in the command line we lied as to what the actual contents of the file are.

<details>
<summary>code to match against the actual yaml (but we won't because it means reparsing every yaml file twice</summary>

```rs
pub fn make_multiline_string_span(
    given_string_lines_span: (usize, usize),
    yaml_text: &str,
    multiline_string_contents: &str,
) -> SourceSpan {
    let col0_of_first_line_of_text = yaml_text
        .char_indices()
        .filter(|a| a.1 == '\n')
        .nth(given_string_lines_span.0 - 2) // -2 = subtract one because span is 1-based
        .map(|a| a.0)                       // subtract one more because we are put
        .expect("to find start of span of test") // at the end of the line
        + 1; // + 1 to go to next line
    let col_of_first_non_space =
        yaml_text[col0_of_first_line_of_text..].chars().position(|x| x != ' ').unwrap();
    let new_line_count = multiline_string_contents.chars().filter(|a| *a == '\n').count();
    let end = yaml_text[col0_of_first_line_of_text + col_of_first_non_space..]
        .char_indices()
        .filter(|a| a.1 == '\n')
        .nth(new_line_count - 1)
        .map(|a| a.0)
        .expect("to find end of span of test");
    (col0_of_first_line_of_text + col_of_first_non_space, end).into()
}
```

Reply:
I added a comment here: [120785a](https://github.com/web-infra-dev/oxc/pull/798/commits/120785aa32b9ac85bbb25662ea19e4c15305ed9e)